core/capabilities/ccip/launcher/deployment.go

```diff
@@ -3,140 +3,48 @@ package launcher
-	}
+type ccipPlugins map[ocrtypes.ConfigDigest]cctypes.CCIPOracle
 
-	return err
+// StartAll will call Oracle.Start on an entire don
+func (c ccipPlugins) StartAll() error {
+	nilPlugins := make(ccipPlugins)
+	return c.Transition(nilPlugins)
 }
 
-// StartActive starts the active OCR instances.
-func (c *ccipDeployment) StartActive() error {
-	var err error
-
-	err = multierr.Append(err, c.commit.active.Start())
-	err = multierr.Append(err, c.exec.active.Start())
-
-	return err
+// CloseAll is used to shut down an entire don immediately
+func (c ccipPlugins) CloseAll() error {
+	nilPlugins := make(ccipPlugins)
+	return nilPlugins.Transition(c)
 }
 
-// CloseActive shuts down the active OCR instances.
-func (c *ccipDeployment) CloseActive() error {
+// Transition manages starting and stopping ocr instances
+// If there are any new config digests, we need to start those instances
+// If any of the previous config digests are no longer present, we need to shut those down
+// We don't care about if they're exec/commit or active/candidate, that all happens in the plugin
+func (c ccipPlugins) Transition(prevPlugins ccipPlugins) error {
 	var err error
 
-	err = multierr.Append(err, c.commit.active.Close())
-	err = multierr.Append(err, c.exec.active.Close())
-
-	return err
-}
-
-// TransitionDeployment handles the active-candidate deployment transition.
-// prevDeployment is the previous deployment state.
-// there are two possible cases:
-//
-// 1. both active and candidate are present in prevDeployment, but only active is present in c.
-// this is a promotion of candidate to active, so we need to shut down the active deployment
-// and make candidate the new active. In this case candidate is already running, so there's no
-// need to start it. However, we need to shut down the active deployment.
-//
-// 2. only active is present in prevDeployment, both active and candidate are present in c.
-// In this case, active is already running, so there's no need to start it. We need to
-// start candidate.
-func (c *ccipDeployment) TransitionDeployment(prevDeployment *ccipDeployment) error {
-	if prevDeployment == nil {
-		return fmt.Errorf("previous deployment is nil")
+	if len(c) > 4 || len(prevPlugins) > 4 {
+		return fmt.Errorf("current ccipPlugins or prevPlugins have more than 4 instances: len(prevPlugins): %d, len(currPlugins): %d", len(prevPlugins), len(c))
 	}
-
-	var err error
-	if prevDeployment.commit.candidate != nil && c.commit.candidate == nil {
-		err = multierr.Append(err, prevDeployment.commit.active.Close())
-	} else if prevDeployment.commit.candidate == nil && c.commit.candidate != nil {
-		err = multierr.Append(err, c.commit.candidate.Start())
-	} else {
-		return fmt.Errorf("invalid active-candidate deployment transition")
+	// This shuts down instances that were present previously, but are no longer needed
+	for digest, oracle := range prevPlugins {
+		if _, ok := c[digest]; !ok {
+			err = multierr.Append(err, oracle.Close())
```

Comment:
Ha interesting this is that this set sub logic is mirrored in the diff code. Maybe needs a lib to unify 🤔

Reply:
https://pkg.go.dev/github.com/smartcontractkit/chainlink-common@v0.3.0/pkg/services#MultiStart
https://pkg.go.dev/github.com/smartcontractkit/chainlink-common@v0.3.0/pkg/services#MultiCloser

Reply:
Updated to start and close in parallel. 

In this situation since we have a max of 4 oracles only, i think using a mapset is unnecessary. We also key the oracles based on config digest, which would be less meaningful in a mapset.